File: Programmers/[64061] 크레인 인형뽑기 게임/SUbbb/64061.java
import java.util.*;

class Solution {
    // 인형 담을 바구니
    Stack<Integer> basket = new Stack<>();
    
    public int solution(int[][] board, int[] moves) {
        int answer = 0;
        
        for (int move : moves) {
            int r = dollIdx(board, move - 1, board.length);
            
            // 인형이 없는 경우, 다음 move 확인
            if (r == -1) continue;
            
            // 바구니의 TOP과 집은 인형 비교, 같다면 바구니 pop, 다르면 집은 인형 push
            int doll = board[r][move - 1];
            if (basket.size() >= 1) {
                if (basket.peek() == doll) { 
                    basket.pop();
                    answer += 2;
                }
                else basket.push(doll);
            } else basket.push(doll);
            
            // 보드 인형 제거
            board[r][move - 1] = 0;
        }
        return answer;
    }
    
    private int dollIdx(int[][] board, int r, int c) {
        // 인형이 위치한 행을 반환하는 함수, 없으면 -1 반환
        for (int i = 0; i < c; i++) 
            if (board[i][r] != 0) return i;
        return -1;
    }
}

File: Programmers/[64061] 크레인 인형뽑기 게임/KingDonggyu/64061.js
function solution(board, moves) {
  let answer = 0;

  const basket = []; // stack
  const top = new Array(board.length + 1).fill(null);

  // 맨 위 인형의 위치 구하기
  for (let y = 0; y < board.length; y++) {
    for (let x = 0; x < board.length; x++) {
      if (board[x][y] === 0) continue;
      top[y + 1] = [x, y];
      break;
    }
  }

  // 인형 뽑기 시작
  moves.forEach((move) => {
    if (!top[move]) return;

    const [x, y] = top[move];
    const doll = board[x][y];

    // 맨 위 인형 위치 업데이트
    top[move] = x + 1 < board.length ? [x + 1, y] : null;

    // 인형 터뜨리기
    if (basket[basket.length - 1] === doll) {
      basket.pop();
      answer += 2;
    } else {
      basket.push(doll);
    }
  });

  return answer;
}
